lib/matplotlib/cm.py

```diff
@@ -367,6 +367,25 @@ def set_array(self, A):
         ----------
         A : ndarray or None
         """
+        if A is None:
+            self._A = None
+            return
+
+        if isinstance(self, mpl.collections.QuadMesh):
+            A_ = A.ravel()
+            width, height = self._meshWidth, self._meshHeight
+            if self._shading in ['flat', 'nearest', 'auto']:
```

Comment:
Sorry, I should have seen this sooner.  This check isn't compatible with `flat` or `auto` - please see the docstring for pcolormesh...

Reply:
I think this function needs to be called when the QuadMesh is initialized to make sure it is compatible with the initialization. I don't know if that means we want to drop extra data for `flat` or not.

Reply:
I'm sorry I don't follow, why isn't the check compatible with `flat` or `auto`?

Reply:
> I think this function needs to be called when the QuadMesh is initialized to make sure it is compatible with the initialization. I don't know if that means we want to drop extra data for `flat` or not.

When quadmesh is initialized the arguments are already validated for faulty data right?

Reply:
Also, when shading = `flat`, should `set_array` only accept data of same shape as X and Y (where extra data is dropped) or should it accept data of where it's dimensions are one less than X and Y, or should it accept both?

Reply:
So the shading for A [NxM]

`nearest`: x[M], y[N]
`flat`: x [M+1], y[N+1] is preferred, but for back-combat x[M], y[N] works by dropping last row and column of A (so it is now [N-1 x M-1]
`auto`: Chooses `nearest` if x [M], y[N], or `flat` if x [M+1], y [N+1]
`gouraud`: x[M+1], y[N+1].

Reply:
Ah, okay thanks for explaining it. 
One more small question I've got is 
https://github.com/matplotlib/matplotlib/blob/81f0624c3baa7a2f4a01975fd9d71199894ab94d/lib/matplotlib/axes/_axes.py#L5727
When shading is set to `auto` and like you've mentioned `auto`: Chooses `nearest` if x [M], y[N], or `flat` if x [M+1], y [N+1].
Shading is set back to being flat when x[M[, y[N] and C is of the same shape, this basically causes `Quadmesh._shading` to be set flat regardless of the `shading` set to `auto`, it's never `nearest` when set to `auto`.